### PR TITLE
Implement `Default` for more types in the standard library

### PR DESCRIPTION
**Pull request by tbu-**
Also add `Hash` to `std::cmp::Ordering` and most possible traits to
`fmt::Error`.

### CODE REVIEW

**Comment by tbu-** [comment]
There was a reddit comment that motivated this (especially for `Mutex` and `RwLock`), but I can't find it anymore.

**Comment by bluss** [comment]
The added traits also should be mentioned in the commit message or PR.

**Comment by tbu-** [comment]
@bluss Updated.

**Comment by alexcrichton** [comment]
Thanks @tbu-! Tagging with `T-libs` to ensure that this comes up during triage.

File: src/libstd/time/duration.rs [hunk 1]
```diff
@@ -188,6 +188,13 @@ impl DivAssign<u32> for Duration {
     }
 }
 
+#[stable(feature = "duration_default", since = "1.9.0")]
+impl Default for Duration {
+    fn default() -> Duration {
+        Duration::new(0, 0)
```
Comment (llogiq) [inline src/libstd/time/duration.rs]: My [question on StackOverflow](https://stackoverflow.com/questions/36477505/why-is-default-not-implemented-for-mutex-rwlock-condvar-duration) has brought up the concern that a 0-duration is *not* a sensible default in most cases. Using this e.g. as a timeout is bound to fail at runtime. It's probably better not to implement Default for Duration at all.

**Comment by llogiq** [comment]
That [reddit comment](https://www.reddit.com/r/rust/comments/4df72s/the_to_string_method_is_now_specialized_for_str/d1sayvd) was by me. This would fix #31865  (and some more) by the way.

I'm also unsure if this should be directly stable.

File: src/libstd/time/duration.rs (hunk 1 above)
Comment (bluss) [inline src/libstd/time/duration.rs]: Why would it fail at runtime? The drawbacks were not fully explained on SO.

Comment (llogiq) [inline src/libstd/time/duration.rs]: The problem is that Durations are often used as Timeouts and unless zero-duration timeouts are special case, the operation will *always* time out by default.

**Comment by llogiq** [comment]
OK, so TIL trait impls are insta-stable. :smile:

Apart from being unsure about the 'correct' default value for Duration (zero seems to make intuitive sense for timestamp arithmetic, but is a bad choice for timeouts), I'm in favor of this. It's more complete than #32807.

File: src/libstd/time/duration.rs (hunk 1 above)
Comment (tbu-) [inline src/libstd/time/duration.rs]: I don't have strong feelings about this, we can definitely remove this if this isn't wanted. Two upsides: Using 0 seems to have some precedence (integers, floats), even though it's not clear where you'll use them. Implementing `Default` for more types means that you can `#[derive]` it in more cases.

Comment (llogiq) [inline src/libstd/time/duration.rs]: Yeah, that was exactly the argument I made in the reddit comment. And of course the default just has to make intuitive sense; we cannot foresee possible applications. Still, I'd like to have more input for this before we commit to an implementation, especially as insta-stability means we cannot easily back off.

**Comment by alexcrichton** [comment]
The libs team discussed this during triage yesterday and the decision was that most of these seem fine but we should avoid doing something just for the sake of doing something. In line with that the default impls for `Path` and `CStr` seem like "just because", so can they be removed for now? Additionally, the `Duration` default seems like it's not "clearly the right choice" so can that be removed for now as well? (erring on the side of being conservative).

Other than that though this looked good to us!

**Comment by tbu-** [comment]
`&CStr` is pretty close to `&str` and `Default` is implemented for that type. `Default` is also implemented for `&[T]` and `&mut [T]`.

The use case I have in mind here is when you have a `Option<&CStr>`, an `unwrap_or_default` should give you the empty string (like for `Option<&str>`).

**Comment by llogiq** [comment]
I also see no problem with `CStr`. `Duration` on the other hand should probably wait.

**Comment by alexcrichton** [comment]
Yeah splitting them out is fine, let's merge this and continue discussion on #32990 

@bors: r+ 3df35a01e9825bbdebb986f980095e468357975f

**Comment by llogiq** [comment]
@tbu- :+1:

**Comment by llogiq** [comment]
This fixes #31865.

**Comment by bombless** [comment]
I cannot think of a more proper default value for `Duration` other than zero though.

IMO if your scenario has a more proper value for a duration it should be a separate type and you may want to express duration itself through a trait.

And the implementation of `Default` might just be used casually, like https://github.com/bombless/to_default/blob/master/examples/simple.rs

**Pull request merged**